test/metabase/driver/postgres_test.clj

```diff
@@ -1218,3 +1218,40 @@
                         (map :schema_name (jdbc/query conn-spec "SELECT schema_name from INFORMATION_SCHEMA.SCHEMATA;"))))
               (is (nil? (some (partial re-matches #"metabase_cache(.*)")
                               (driver/syncable-schemas driver/*driver* (mt/db))))))))))))
+
+(deftest table-privileges-test
+  (mt/test-driver :postgres
+    (testing "`table-privileges` should return the correct data for current_user and role privileges"
+      (mt/with-empty-db
+        (let [conn-spec (sql-jdbc.conn/db->pooled-connection-spec (mt/db))
+              get-privileges (fn []
+                               (sql-jdbc.conn/with-connection-spec-for-testing-connection [spec [:postgres (assoc (:details (mt/db))
+                                                                                                                  :user "privilege_rows_test_example_role")]]
+                                 (with-redefs [sql-jdbc.conn/db->pooled-connection-spec (fn [_] spec)]
+                                   (driver/current-user-table-privileges driver/*driver* (mt/db)))))]
+          (try
+            (jdbc/execute! conn-spec (str "CREATE SCHEMA foo;"
+                                          "CREATE TABLE foo.bar (id INTEGER);"
+                                          "CREATE TABLE foo.baz (id INTEGER);"
+                                          "CREATE ROLE privilege_rows_test_example_role WITH LOGIN;"
+                                          "GRANT SELECT ON foo.baz TO privilege_rows_test_example_role;"
+                                          "GRANT UPDATE ON foo.baz TO privilege_rows_test_example_role;"))
+            (testing "check that without USAGE privileges, nothing is returned"
+              (is (= []
+                     (get-privileges))))
+            (testing "with USAGE privileges, select is returned"
+              (jdbc/execute! conn-spec "GRANT USAGE ON SCHEMA foo TO privilege_rows_test_example_role;")
+              (is (= [{:is_current_user true,
+                       :role "privilege_rows_test_example_role",
+                       :schema "foo",
+                       :table "baz",
+                       :select true,
+                       :update true,
+                       :insert false,
+                       :delete false}]
+                     (get-privileges))))
+            (finally
+              (doseq [stmt ["REVOKE ALL PRIVILEGES ON TABLE foo.baz FROM privilege_rows_test_example_role;"
+                            "REVOKE ALL PRIVILEGES ON SCHEMA foo FROM privilege_rows_test_example_role;"
+                            "DROP ROLE privilege_rows_test_example_role;"]]
+                (jdbc/execute! conn-spec stmt)))))))))
```

Comment:
is it better to run the test in a transaction and rollback at the end?

Reply:
That won't allow me to switch users and check the privileges as I'm doing here
https://github.com/metabase/metabase/blob/ef3a4480e1bb09ed138ecf4a2887ac57520a4f5f/test/metabase/driver/postgres_test.clj#L1227-L1231

Reply:
Sorry I didn't see your last comment. The DB is deleted but created role isn't deleted when the DB is deleted, only the privileges are. We could have a `finally` block that executes "DROP ROLE privilege_rows_test_example_role;" outside the `mt/with-empty-db` expression, but there could be an error thrown inside the body of `mt/with-empty-db` or `sql-jdbc.conn/db->pooled-connection-spec` and in those cases we wouldn't want to run the `finally` block, because the role wouldn't have been created. I think it's fine to leave it here.